great_expectations/experimental/datasources/interfaces.py

```diff
@@ -371,94 +380,88 @@ def execution_engine_type(self) -> Type[ExecutionEngine]:
-class Batch:
-    # Instance variable declarations
-    _datasource: Datasource
-    _data_asset: DataAsset
-    _batch_request: BatchRequest
-    _data: BatchDataType
-    _id: str
+class Batch(ExperimentalBaseModel):
+    """This represents a batch of data.
+
+    This is usually not the data itself but a hook to the data on an external datastore such as
+    a spark or a sql database. An exception exists for pandas or any in-memory datastore.
+    """
+
+    datasource: Datasource
+    data_asset: DataAsset
+    batch_request: BatchRequest
+    data: "BatchData"
+    id: str = ""
     # metadata is any arbitrary data one wants to associate with a batch. GX will add arbitrary metadata
     # to a batch so developers may want to namespace any custom metadata they add.
-    metadata: Dict[str, Any]
+    metadata: Dict[str, Any] = {}
 
     # TODO: These legacy fields are currently required. They are only used in usage stats so we
     #       should figure out a better way to anonymize and delete them.
-    _legacy_batch_markers: BatchMarkers
-    _legacy_batch_spec: BatchSpec
-    _legacy_batch_definition: BatchDefinition
+    batch_markers: "BatchMarkers" = Field(..., alias="legacy_batch_markers")
+    batch_spec: BatchSpec = Field(..., alias="legacy_batch_spec")
+    batch_definition: "BatchDefinition" = Field(..., alias="legacy_batch_definition")
 
-    def __init__(
-        self,
-        datasource: Datasource,
-        data_asset: DataAsset,
-        batch_request: BatchRequest,
-        # BatchDataType is Union[core.batch.BatchData, pd.DataFrame, SparkDataFrame].  core.batch.Batchdata is the
-        # implicit interface that Datasource implementers can use. We can make this explicit if needed.
-        data: BatchDataType,
-        # Legacy values that should be removed in the future.
-        legacy_batch_markers: BatchMarkers,
-        legacy_batch_spec: BatchSpec,
-        legacy_batch_definition: BatchDefinition,
-        # Optional arguments
-        metadata: Optional[Dict[str, Any]] = None,
-    ) -> None:
-        """This represents a batch of data.
-
-        This is usually not the data itself but a hook to the data on an external datastore such as
-        a spark or a sql database. An exception exists for pandas or any in-memory datastore.
-        """
-        # These properties are intended to be READ-ONLY
-        self._datasource: Datasource = datasource
-        self._data_asset: DataAsset = data_asset
-        self._batch_request: BatchRequest = batch_request
-        self._data: BatchDataType = data
-        self.metadata = metadata or {}
-
-        self._legacy_batch_markers = legacy_batch_markers
-        self._legacy_batch_spec = legacy_batch_spec
-        self._legacy_batch_definition = legacy_batch_definition
-
-        # computed property
-        # We need to unique identifier. This will likely change as I get more input
+    class Config:
+        allow_mutation = False
+        arbitrary_types_allowed = True
```

Comment:
Why do we need to set `arbitrary_types_allowed = True`?

Reply:
Because `pydantic` only has out-of-the-box validations for [certain common types](https://docs.pydantic.dev/usage/types/#standard-library-types). This option enables an `isinstance` check for our custom types. More detailed [validations can be written for our types](https://docs.pydantic.dev/usage/types/#custom-data-types) in the future.

[https://docs.pydantic.dev/usage/types/#arbitrary-types-allowed](https://docs.pydantic.dev/usage/types/#arbitrary-types-allowed)